### DISCUSSION TITLE
libtorch jit forward error

### DISCUSSION THREAD

**Discussion by moodeerf** [snakers4/silero-vad#221] [Q&A] [answered]
### Discussed in https://github.com/snakers4/silero-vad/discussions/154
Hi, I had encounted the same problem.
When I call the **'forward'** function with one parameters (_like this forward(inputs)_), the code can be compiled correctly, but it can't run with the error '_what(): forward() is missing value for argument 'sr'._'.
However, When I call the **'forward'** function with two parameters (_like this forward(inputs,sr)_), the code can't be compiled correctly with the error '_candidate expects 1 argument, 2 provided_'. 
I doubt that the jit model was not  exported correctly, please check it at your convenience. By the way, I can call the **'forward'** function with one parameters (_like this forward(inputs)_) with the old jit model.
On the other hand, the function **'reset_states'** of the jit model can't be used in C++ code, can you provide a name so we can use it like this _'model.get_method('name')()' or 'model.run_method('name')'_

**Comment by snakers4** [comment]
Hi,

Not sure that we will be opening old models, but we are planning a major update some time soon.
This may be a good time for debugging with libtorch as well.
Maybe this is an issue of libtorch itself, idk.

> with the old jit model.

If remember correctly, the old jit models had a large chunk size, did not have an SR parameter, did not have internal state.

> On the other hand, the function 'reset_states' of the jit model can't be used in C++ code, can you provide a name so we can use it like this 'model.get_method('name')()' or 'model.run_method('name')'

It is the case for all C++ models, only `.forward()` can be invoked, or is it a particular error with this model?

**Comment by moodeerf** [comment]
> Hi,
> 
> Not sure that we will be opening old models, but we are planning a major update some time soon. This may be a good time for debugging with libtorch as well. Maybe this is an issue of libtorch itself, idk.
> 
> > with the old jit model.
> 
> If remember correctly, the old jit models had a large chunk size, did not have an SR parameter, did not have internal state.
> 
> > On the other hand, the function 'reset_states' of the jit model can't be used in C++ code, can you provide a name so we can use it like this 'model.get_method('name')()' or 'model.run_method('name')'
> 
> It is the case for all C++ models, only `.forward()` can be invoked, or is it a particular error with this model?

Em...,
Firstly I had used other jit models，**.forward()** can be invoked correctly , so most likely not the issue of libtorch itself, idk.
Secondly, If **'reset_states'** is able to be invoked in C++ code, **'reset_states'** need be defined in Python code follow the rules of _torch.jit.export_ [https://pytorch.org/docs/stable/jit.html](https://github.com/snakers4/silero-vad/issues/url)
For example:
the function **' right_context'** is defined in line:549 of the file
[https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py](https://github.com/snakers4/silero-vad/issues/url)
and the function can be invoked in the line:58 of the file [https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/torch_asr_model.cc](https://github.com/snakers4/silero-vad/issues/url)
So,the **'reset_states'** should be defined like above.

**Comment by snakers4** [comment]
> Firstly I had used other jit models，.forward() can be invoked correctly , so most likely not the issue of libtorch itself, idk.

With multiple input variables?

**Comment by moodeerf** [comment]
> 

I will check it .

**Comment by moodeerf** [comment, marked as answer]
> > Firstly I had used other jit models，.forward() can be invoked correctly , so most likely not the issue of libtorch itself, idk.
> 
> With multiple input variables?
 Hi,
Finally i solved the problem like this .
`std::vector<torch::jit::IValue> inputs;`
 `inputs.push_back(torch::from_blob(buffer,{chunk_size},torch::kFloat32));`
`int sr = 16000;`
`inputs.push_back(sr);`
`at::Tensor output = module.forward(inputs).toTensor();`
